File: src/main/java/uk/co/ben_gibson/git/link/Url/Factory/UrlFactoryLocator.java
package uk.co.ben_gibson.git.link.Url.Factory;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.project.Project;
import uk.co.ben_gibson.git.link.Preferences;
import uk.co.ben_gibson.git.link.Url.Factory.Exception.UrlFactoryException;

import java.util.ArrayList;
import java.util.List;

public class UrlFactoryLocator
{
    private List<UrlFactory> factories = new ArrayList<>();
    private Preferences preferences;

    public static UrlFactoryLocator getInstance(Project project) {
        return project.getService(UrlFactoryLocator.class);
    }

    public UrlFactoryLocator(Project project) {

        this.preferences = Preferences.getInstance(project);

        this.factories.add(ApplicationManager.getApplication().getService(GitLabUrlFactory.class));
        this.factories.add(ApplicationManager.getApplication().getService(GitHubUrlFactory.class));
        this.factories.add(ApplicationManager.getApplication().getService(GogsUrlFactory.class));
        this.factories.add(ApplicationManager.getApplication().getService(BitbucketServerUrlFactory.class));
        this.factories.add(ApplicationManager.getApplication().getService(BitbucketCloudUrlFactory.class));
        this.factories.add(project.getService(CustomUrlFactory.class));
    }

    public UrlFactory locate() throws UrlFactoryException {
        for (UrlFactory factory : this.factories) {
            if (factory.supports(this.preferences.getRemoteHost())) {
                return factory;
            }
        }

        throw UrlFactoryException.unsupportedRemoteHost(this.preferences.getRemoteHost());
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/url/factory/UrlFactoryLocator.kt
package uk.co.ben_gibson.git.link.url.factory

import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import uk.co.ben_gibson.git.link.platform.*

@Service
class UrlFactoryLocator {
    fun locate(platform: Platform) : UrlFactory {
        return when(platform) {
            is Azure -> service<AzureUrlFactory>()
            is BitbucketServer -> service<BitbucketServerUrlFactory>()
            is TemplatedPlatform -> TemplatedUrlFactory(platform.templates)
        }
    }
}
